src/main/scala/firrtl/ir/IR.scala

```diff
@@ -211,8 +211,35 @@ abstract class Width extends FirrtlNode {
   }
 }
 /** Positive Integer Bit Width of a [[GroundType]] */
-case class IntWidth(width: BigInt) extends Width {
+object IntWidth {
+  private val maxCached = 1024
+  private val cache = new Array[IntWidth](maxCached + 1)
+  def apply(width: BigInt): IntWidth = {
+    if (0 <= width && width <= maxCached) {
+      val i = width.toInt
+      var w = cache(i)
+      if (w eq null) {
+        w = new IntWidth(width)
+        cache(i) = w
+      }
+      w
+    } else new IntWidth(width)
+  }
+  // For pattern matching
+  def unapply(w: IntWidth): Option[BigInt] = Some(w.width)
+}
+class IntWidth(val width: BigInt) extends Width with Product {
   def serialize: String = s"<$width>"
+  override def equals(that: Any) = that match {
+    case w: IntWidth => width == w.width
+    case _ => false
+  }
+  override def hashCode = width.toInt
+  override def toString = s"IntWidth($width)"
+  def copy(width: BigInt = width) = new IntWidth(width)
+  def canEqual(that: Any) = that.isInstanceOf[Width]
+  def productArity = 1
+  def productElement(int: Int) = int match { case 0 => width }
```

Comment:
The copy method should delegate to the object factory method so you benefit from caching.

You should also ' override def productPrefix = "IntWidth" ' and consider using that in the toString implementation (although IntWidth likely isn't meant to be subclasses so not as critical).
productElement should throw an IndexOutOfBoundsException instead of a MatchError.

I'm consulting with some scala colleagues on the semantics on marking a case class as 'abstract'. It seems like it only suppresses the factory method and copy method, which you can just re-add yourself. (You will need to append {} to the new operator since you can't instantiate an abstract class so you need to 'refine' the class to a concrete class.) It is unclear whether this behavior is reliable enough to be used.